File: actors/adder/macro.f90
program macro
    use ymmsl
    use libmuscle
    implicit none

    real (selected_real_kind(15)) :: t_cur, t_next, t_max, dt
    real (selected_real_kind(15)) :: a, b, s

    type(LIBMUSCLE_PortsDescription) :: ports
    type(LIBMUSCLE_Instance) :: instance

    type(LIBMUSCLE_Message) :: rmsg
    type(LIBMUSCLE_DataConstRef) :: rdata

    type(LIBMUSCLE_Message) :: smsg
    type(LIBMUSCLE_Data) :: sdata

    write(*,*) "Starting adder Fortran macro actor"
    
    ports = LIBMUSCLE_PortsDescription_create()
    call LIBMUSCLE_PortsDescription_add(ports, YMMSL_OPERATOR_O_I, 'state_out')
    call LIBMUSCLE_PortsDescription_add(ports, YMMSL_OPERATOR_S, 'state_in')
    instance = LIBMUSCLE_Instance_create(ports)
    call LIBMUSCLE_PortsDescription_free(ports)

    do while (LIBMUSCLE_Instance_reuse_instance(instance))
        ! F_INIT
        t_max = LIBMUSCLE_Instance_get_setting_as_real8(instance, 't_max')
        dt = LIBMUSCLE_Instance_get_setting_as_real8(instance, 'dt')

        t_cur = 0.0
        do while (t_cur + dt <= t_max)
            ! O_I
            t_next = t_cur + dt

            ! Send two random numbers to the micro model
            call random_number(a)
            call random_number(b)

            sdata = LIBMUSCLE_Data_create_nils(2_LIBMUSCLE_size)
            call LIBMUSCLE_Data_set_item(sdata, 1_LIBMUSCLE_size, a)
            call LIBMUSCLE_Data_set_item(sdata, 2_LIBMUSCLE_size, b)
  
            smsg = LIBMUSCLE_Message_create(t_cur, sdata)
            if (t_next + dt <= t_max) then
                call LIBMUSCLE_Message_set_next_timestamp(smsg, t_next)
            end if
            call LIBMUSCLE_Instance_send(instance, 'state_out', smsg)
            call LIBMUSCLE_Message_free(smsg)
            call LIBMUSCLE_Data_free(sdata)

            ! S
            ! Receive the sum from the micro model
            rmsg = LIBMUSCLE_Instance_receive(instance, 'state_in')
            rdata = LIBMUSCLE_Message_get_data(rmsg)

            s = LIBMUSCLE_DataConstRef_as_real8(rdata)
            call LIBMUSCLE_DataConstRef_free(rdata)
            if (abs(a + b - s) > 1e-10) then
                call LIBMUSCLE_Instance_error_shutdown(instance, "Received unexpected message")
                call exit(1)
            end if

            t_cur = t_cur + dt
        end do

    end do

end program macro

File: actors/adder/macro.cpp
// MUSCLE3 actor requesting a micro model to add two numbers

#include <cstdlib>
#include <iostream>

// MUSCLE imports
#include <libmuscle/libmuscle.hpp>
#include <ymmsl/ymmsl.hpp>

int main(int argc, char *argv[]) {
  std::cout << "Starting adder C++ macro actor" << std::endl;
  libmuscle::Instance instance(argc, argv,
                               {
                                   {ymmsl::Operator::O_I, {"state_out"}},
                                   {ymmsl::Operator::S, {"state_in"}},
                               });

  while (instance.reuse_instance()) {
    // F_INIT
    double t_max = instance.get_setting_as<double>("t_max");
    double dt = instance.get_setting_as<double>("dt");

    double t_cur = 0.0;
    while (t_cur + dt <= t_max) {
      // O_I
      double t_next = t_cur + dt;

      // Send two random numbers to the micro model
      double a = (double)std::rand() / RAND_MAX;
      double b = (double)std::rand() / RAND_MAX;
      libmuscle::Message msg(t_cur, libmuscle::Data::list(a, b));
      if (t_next + dt <= t_max) msg.set_next_timestamp(t_next);
      instance.send("state_out", msg);

      // S
      // Receive the sum from the micro model
      msg = instance.receive("state_in");
      double s = msg.data().as<double>();
      if (std::abs(a + b - s) > 1e-10) {
        instance.error_shutdown("Received unexpected message");
        return 1;
      }

      t_cur += dt;
    }
  }
}
